model/flow/identity.go

```diff
@@ -394,6 +394,9 @@ func (iy *Identity) UnmarshalMsgpack(b []byte) error {
 }
 
 func (iy *IdentitySkeleton) EqualTo(other *IdentitySkeleton) bool {
+	if other == nil {
+		return false
+	}
```

Comment:
I assume this is coming in from `master`, but I'm curious why this was added. What if both `iy` and `other` are nil?

Reply:
not sure, I thought maybe some cases due to new crypto, but I didn't check, seemed not harmful so I've left it. I will also try if something breaks without it. @tarakby have you added this by any chance?

Reply:
I think we really need @tarakby's input here.

* Reviewing `flow.Identity`, there are two methods for checking equality: `Equals`  and `EqualTo`, where it seems @tarakby added the former in May
* Neither method has documentation and their naming indicates that both do very similar things. This lack of documentation should be fixed (happy to roll these few lines into our PR)
   - based on the implementation, `Equals`  seems to handle nil inputs (but not a nil receiver), while `EqualTo` does not handle a nil input. Is that correct? 
   - `Equals` calls the `PublicKey.Equals` method for the keys. However, the [PublicKey API](https://github.com/onflow/flow-go/blob/456c1318b00b9131ecfe8d95a0813e458ebe7fb1/crypto/sign.go#L226-L229) does not specify behaviour for nil inputs nor nil receivers. I think we need to add documentation. Does the method handle nil receivers, because `EqualTo` tolerates nil keys? 
   - To me, the most consistent convention would be to consider two nil objects identical (principle of [Vacuous truth](https://en.wikipedia.org/wiki/Vacuous_truth)), but I think neither implementation `Equals`  or `EqualTo`  adheres to this principle.

Reply:
Just checked and at least in `master`, `Equals` is only used in tests. All test still pass, when replacing `Equals` -> `EqualTo`

Reply:
Yes I added it in this commit https://github.com/onflow/flow-go/commit/a4fb4357656dd6fbbc5e101cda001471a3445c81 to fix tests (see commit code) that have been testing equality of identities (and their public key fields) based on struct equalities. That was not right because comparing public keys should be based on the `Equals` method (two public keys may be equal but have different structs in memory).

I added the new function only to fix the tests and I didn't pay attention to the existing `EqualTo` that seems to compare public keys properly. One of the two functions should be enough, the new one I added is redundant. 

Regarding the `nil` cases, the function I added makes the assumption that `iy` isn't `nil`, though this is implicit and not documented (not great my bad). Feel free to clean up the functions and the `nil` case treatment.
Another poor documented function is the public key comparison within the crypto module, it assumes the receiver and argument are both non nil. I will update the crypto module to either address the nil cases or document the assumptions clearly.

Reply:
( sorry @AlexHentschel I just noticed that I replied without seeing your comments as my opened the conversation early and didn't refresh it before sending my reply )

Reply:
> Just checked and at least in master, Equals is only used in tests 

That's correct, as you see on the commit https://github.com/onflow/flow-go/commit/a4fb4357656dd6fbbc5e101cda001471a3445c81, I only used that function to fix the tests. I should have even added the new function as a test utility.

Reply:
opened this issue: https://github.com/onflow/crypto/issues/4

Reply:
Agree with not supporting nil `*Identity`s 👍